internal/resource/resource_service.go

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package resource
+
+import (
+	"sync"
+
+	"github.com/nginx/agent/v3/api/grpc/mpi/v1"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
+//counterfeiter:generate . resourceServiceInterface
+type resourceServiceInterface interface {
+	AddInstance(instanceList []*v1.Instance) *v1.Resource
+	UpdateInstance(instanceList []*v1.Instance) *v1.Resource
+	DeleteInstance(instanceList []*v1.Instance) *v1.Resource
+}
+
+type ResourceService struct {
+	resource      *v1.Resource
+	resourceMutex sync.Mutex
+}
+
+func NewResourceService() *ResourceService {
+	return &ResourceService{
+		resource: &v1.Resource{
+			Instances: []*v1.Instance{},
+		},
+		resourceMutex: sync.Mutex{},
+	}
+}
+
+func (r *ResourceService) AddInstance(instanceList []*v1.Instance) *v1.Resource {
+	r.resourceMutex.Lock()
```

Comment:
I'd suggest using `defer` here to unlock the mutex. That way the `return r.resource` will also be locked.
```suggestion
	r.resourceMutex.Lock()
	defer r.resourceMutex.Unlock()
```